Feature/formatter codesprint

davidblasby opened the pull request:
# FormatterApi and integration in OGCAPIRec

## Description
This PR adds a flexible metadata-formatting framework to GN5. It provides components that implement a FormatterAPI, which can be used programmatically in code or accessed through service endpoints. The FormatterAPI has been designed to be more versatile than its GN4 counterpart, incorporating feedback from previous PRs and discussions.
This new system makes it easier to define formatters that don't depend on GN4's Schema infrastructure (which is only outlined here and will be detailed in future work). For example, the included IndexFormatter demonstrates how to create formatters entirely in Java using the Elasticsearch index (via IndexFormatterProcessor).
The PR also significantly improves the OGC API Records implementation, particularly in format negotiation, by integrating the FormatterAPI.

## Type of Change
Select one:
- [ ] Bug fix
- [x] New feature
- [x] Documentation update
- [ ] Refactoring

## Related Issue
Closes/updates these PR #78, #94, #96
Updates these issues #50, #66, #72, #24

## Approach
- [Formatters and FormatterAPI design](https://github.com/GeoCat/geonetwork/blob/feature/formatter-codesprint/docs/manual/docs/Developer/formatters.md) 
- OGC API Records improvements and format negotiation
- OGC API Records FormatterAPI integration

## How to Verify
1. Run GN5 with GN4 as you prefer (or you can use [this guide](https://github.com/geonetwork/geonetwork/blob/main/docs/manual/docs/GN4-Integration/index.md))
2. Put some ISO19139 metadata in GN database and index (you can create metadata using GN4) 
3. You have two endpoints to test formatters:
 a. Formatters API http://localhost:7979/api/records/_metadataid_/formatters/_someformatter_
 b. OGC API Records http://localhost:7979/ogcapi-records/collections/_catalogueId_/items/_metadataid_ using the alternate formats URL listed in the default JSON output

## Checklist
- [x] Code is formatted and linted
- [x] All tests pass locally and in CI
- [x] Tests updated (if applicable)
- [x] Documentation updated (if applicable)

## Credits
This PR builds upon and replaces the work done in:
- PR #78 - API / Formatter migration check
- PR #94 - API / Formatter migration improvements 
- PR #96 - FormatterApi

Special thanks to the contributors of the previous PRs @fxprunayre @josegar74 @davidblasby

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/geonetwork/geonetwork?pullRequest=97) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/geonetwork/geonetwork?pullRequest=97) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: david-blasby<br/>:x: fxprunayre<br/>:x: josegar74<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/geonetwork/geonetwork?pullRequest=97) it.</sub>

davidblasby commented:
looks like @fxprunayre  and @josegar74  haven't filled in the CLA.

davidblasby commented:
@jahow @fgravin - regarding our discussion on CSW-like output.

I've added a CSW-xml output format for record lists (`/collections/<id>/items` queries). 

This isn't anything fancy - it just puts the CSW required tags in the output and puts the raw (GN DB table `metadata` - `data` column) inside.  I think this will work for the majority of XML-metadata records and fit in with normal CSW processing.

PascalLike commented on GeoCat/Jenkinsfile:
TODO: Remove this before merging PR

davidblasby commented:
@PascalLike - thank for updates!

davidblasby commented:
original summary text:

added this PR so we can put comments in (and it will eventually merge with main).

rebased with the minor changes from main
minor change to how @Cachable is working
switched from static to dynamic cachemanager
When using cacheManager.setCacheNames(List.of("mycache")); it switches to static mode and you cannot create new caches with the @Cachable annotations.

Added a 5 minute cache in the AuthenticationFacade that gets the the user's group/profile (just for speed). Likely not needed.
NOTE: the AuthenticationFacade was already in the security shared module. It needs some review/rework, but should be ok for now.
